theano/gpuarray/GpuMaxAndArgmax.py

```diff
@@ -0,0 +1,209 @@
+            unsigned* %(name)s_axes_to_reduce = new unsigned[%(name)s_redux_len];
+            for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
+                %(name)s_axes_to_reduce[i] =
+                    (unsigned)( ((%(axes_ctype)s*)PyArray_DATA(%(axes)s)) [i * (PyArray_STRIDES(%(axes)s)[0] / sizeof(%(axes_ctype)s))] );
+            }
+            """
+        ret += """
+        size_t  %(name)s_output_ndim = %(name)s_input_ndim - %(name)s_redux_len;
+        size_t* %(name)s_output_dims = NULL;
+        if(%(name)s_output_ndim == 0) {
+            /* Current backend function GpuArray_maxandargmax does not work when
+             * all axes need to be reduced. So to handle this case, we create a view
+             * of the input as a matrix with 1 row and as many columns as elements
+             * in the input, so that the 2nd dimenson of the matrix will be reduced. */
+            size_t total_size = 1;
+            for(size_t i = 0; i < %(name)s_input_ndim; ++i) {
+                total_size *= PyGpuArray_DIM(%(X)s, i);
+            }
+            size_t newdims[2] = {1, total_size};
+            %(name)s_input = new GpuArray;
+            if(GA_NO_ERROR !=
+                GpuArray_reshape(%(name)s_input, &%(X)s->ga, 2, newdims, GA_ANY_ORDER, 0)
+            ) {
+                %(fail)s
+            }
+            %(name)s_redux_len = 1;
+            %(name)s_axes_to_reduce[0] = 1;
+        } else {
+            %(name)s_output_dims = new size_t[%(name)s_output_ndim];
+            if(%(name)s_redux_len == 1) {
+                for(unsigned i = 0; i < %(name)s_axes_to_reduce[0]; ++i) {
+                    %(name)s_output_dims[i] = PyGpuArray_DIM(%(X)s, i);
+                }
+                for(unsigned i = %(name)s_axes_to_reduce[0] + 1; i < %(name)s_input_ndim; ++i) {
+                    %(name)s_output_dims[i-1] = PyGpuArray_DIM(%(X)s, i);
+                }
+            } else {
+                int64_t current_input_pos = -1;
+                int64_t current_output_pos = -1;
+                for(unsigned i = 0; i < %(name)s_redux_len; ++i) {
+                    for(++current_input_pos; current_input_pos < %(name)s_axes_to_reduce[i]; ++current_input_pos) {
+                        %(name)s_output_dims[++current_output_pos] = PyGpuArray_DIM(%(X)s, current_input_pos);
+                    }
+                }
+                for(++current_input_pos; current_input_pos < %(name)s_input_ndim; ++current_input_pos) {
+                    %(name)s_output_dims[++current_output_pos] = PyGpuArray_DIM(%(X)s, current_input_pos);
+                }
+            }
+        }
+        if(theano_prep_output(&%(max)s, %(name)s_output_ndim, %(name)s_output_dims, %(max_typecode)s, GA_C_ORDER, %(X)s->context)) {
+            %(fail)s
+        }
+        if(theano_prep_output(&%(argmax)s, %(name)s_output_ndim, %(name)s_output_dims, %(argmax_typecode)s, GA_C_ORDER, %(X)s->context)) {
```

Comment:
It seems there is no GA_INT64 in libgpuarray doc: http://deeplearning.net/software/libgpuarray/c_api.html

Nor in types.h : https://github.com/Theano/libgpuarray/blob/master/src/gpuarray/types.h

I think I should better let pygpu.gpuarray.dtype_to_typecode do the work, for a future compatibility !